pkg/user/tx_client_test.go

```diff
@@ -208,11 +245,62 @@ func (suite *TxClientTestSuite) TestConfirmTx() {
 		msg := bank.NewMsgSend(addr, testnode.RandomAddress().(sdk.AccAddress), sdk.NewCoins(sdk.NewInt64Coin(app.BondDenom, 1+balance)))
 		resp, err := suite.txClient.BroadcastTx(suite.ctx.GoContext(), []sdk.Msg{msg}, fee, gas)
 		require.NoError(t, err)
-		require.NotNil(t, resp)
-		_, err = suite.txClient.ConfirmTx(suite.ctx.GoContext(), resp.TxHash)
+		require.Equal(t, resp.Code, abci.CodeTypeOK)
+
+		nonce, signer, exists := suite.txClient.GetTxFromLocalMempool(resp.TxHash)
+		require.True(t, exists)
+		require.Equal(t, suite.txClient.DefaultAccountName(), signer)
+		seq := suite.txClient.Signer().Account(suite.txClient.DefaultAccountName()).Sequence()
+		// Successfully broadcast transaction increases the nonce
+		// txInfo is indexed before the nonce is increased
+		require.Equal(t, seq, nonce+1)
+
+		confirmTxResp, err := suite.txClient.ConfirmTx(suite.ctx.GoContext(), resp.TxHash)
 		require.Error(t, err)
+		require.Nil(t, confirmTxResp)
 		code := err.(*user.ExecutionError).Code
 		require.NotEqual(t, abci.CodeTypeOK, code)
+		nonce, signer, exists = suite.txClient.GetTxFromLocalMempool(resp.TxHash)
+		require.False(t, exists)
+		require.Zero(t, nonce)
+		require.Zero(t, signer)
+	})
+
+	t.Run("should adjust nonce when evicted", func(t *testing.T) {
+		// Set ttl to 1ns to evict tx immediately
+		suite.testnodeConfig.TmConfig.Mempool.TTLDuration = 1 * time.Nanosecond
+
+		// Keep submitting the transaction until we get the eviction error
+		sender := suite.txClient.Signer().Account(suite.txClient.DefaultAccountName())
+		msg := bank.NewMsgSend(sender.Address(), testnode.RandomAddress().(sdk.AccAddress), sdk.NewCoins(sdk.NewInt64Coin(app.BondDenom, 10)))
+		var seqBeforeEviction uint64
+		for {
+			// There is a chance one of the txs will get into a block so we have to keep updating the sequence
+			seqBeforeEviction = sender.Sequence()
+			resp, err := suite.txClient.BroadcastTx(suite.ctx.GoContext(), []sdk.Msg{msg}, fee, gas)
+			require.NoError(t, err)
+			_, err = suite.txClient.ConfirmTx(suite.ctx.GoContext(), resp.TxHash)
+			if err != nil {
+				if err.Error() == "tx was evicted from the mempool" {
```

Comment:
if there's a bug, could this just loop indefinitely?

Reply:
if for some reason this loops indefinitely(which should not be the case) the test will hang until it times out so we'll know that the loop is broken

Reply:
not blocking, but what would be a way to avoid the test taking 10 minutes to fail?

Reply:
I can limit the number of loops to 5 for example. It will fail before then.